flink-clients/src/main/java/org/apache/flink/client/program/StreamContextEnvironment.java

```diff
@@ -269,20 +274,86 @@ private List<String> collectNotAllowedConfigurations() {
                                         ConfigurationNotAllowedMessage.ofConfigurationChange(
                                                 k, v)));
 
-        if (!Arrays.equals(originalCheckpointConfigSerialized, serializeConfig(checkpointCfg))) {
+        final Configuration enrichedClusterConfig = new Configuration(originalClusterConfig);
+        enrichProgramConfigWildcards(enrichedClusterConfig);
+
+        // Check CheckpointConfig
+        final CheckpointConfig clusterCheckpointConfig = new CheckpointConfig();
+        clusterCheckpointConfig.configure(enrichedClusterConfig);
+        if (!Arrays.equals(
+                serializeConfig(clusterCheckpointConfig), serializeConfig(checkpointCfg))) {
             errors.add(
                     ConfigurationNotAllowedMessage.ofConfigurationObject(
                             checkpointCfg.getClass().getSimpleName()));
         }
 
-        if (!Arrays.equals(originalExecutionConfigSerialized, serializeConfig(config))) {
+        // Check ExecutionConfig
+        final ExecutionConfig clusterExecutionConfig = new ExecutionConfig();
+        clusterExecutionConfig.configure(enrichedClusterConfig, this.getUserClassloader());
+        if (!Arrays.equals(serializeConfig(clusterExecutionConfig), serializeConfig(config))) {
             errors.add(
                     ConfigurationNotAllowedMessage.ofConfigurationObject(
                             config.getClass().getSimpleName()));
         }
+
         return errors;
     }
 
+    private static final class WildcardOption<T> {
+        private final ConfigOption<T> option;
+        private final Function<StreamContextEnvironment, T> getter;
+
+        WildcardOption(ConfigOption<T> option, Function<StreamContextEnvironment, T> getter) {
+            this.option = option;
+            this.getter = getter;
+        }
+
+        void enrich(Configuration mutableConfig, StreamContextEnvironment fromEnv) {
+            mutableConfig.set(option, getter.apply(fromEnv));
+        }
+
+        void remove(Configuration mutableConfig) {
+            mutableConfig.removeConfig(option);
+        }
+    }
+
+    private static final Map<String, WildcardOption<?>> supportedProgramConfigWildcards =
+            new HashMap<>();
+
+    static {
+        supportedProgramConfigWildcards.put(
+                PipelineOptions.GLOBAL_JOB_PARAMETERS.key(),
+                new WildcardOption<>(
+                        PipelineOptions.GLOBAL_JOB_PARAMETERS,
+                        env -> env.getConfig().getGlobalJobParameters().toMap()));
+    }
+
+    private void enrichProgramConfigWildcards(Configuration mutableConfig) {
+        for (String key : programConfigWildcards) {
+            final WildcardOption<?> option = supportedProgramConfigWildcards.get(key);
```

Comment:
Why do we need to limit the "wildcard" configurations at all? This should be up to the cluster admin to decide.

We should ensure to remove wildcards defined in the `clusterConfiguration` to prevent users from overriding it.

Reply:
Your comment makes absolutely sense. But the implementation is a result of the terrible configuration story that we have to support right now. This is also the reason why I added a comment to `ExecutionConfig` and `CheckpointCfg`. If everything would be backed by `Configuration` instead of the POJOs that partially materialize the config options, the implementation could be way easier. Currently, we need to manually provide the backward path from POJO to ConfigOption value manually.